presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DereferencePushDownJoin.java

```diff
@@ -0,0 +1,116 @@
+
+import com.facebook.presto.metadata.Metadata;
+import com.facebook.presto.sql.parser.SqlParser;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.iterative.Rule.Context;
+import com.facebook.presto.sql.planner.plan.Assignments;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.ProjectNode;
+import com.facebook.presto.sql.tree.ComparisonExpression;
+import com.facebook.presto.sql.tree.DereferenceExpression;
+import com.facebook.presto.sql.tree.Expression;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.facebook.presto.sql.planner.SymbolsExtractor.extractAll;
+import static com.facebook.presto.sql.planner.plan.JoinNode.EquiJoinClause;
+import static com.facebook.presto.sql.planner.plan.Patterns.join;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static com.google.common.collect.Iterables.getOnlyElement;
+
+public class DereferencePushDownJoin
+        extends DereferencePushDownRule<JoinNode>
+{
+    public DereferencePushDownJoin(Metadata metadata, SqlParser sqlParser)
+    {
+        super(metadata, sqlParser, join());
+    }
+
+    @Override
+    protected PlanNode dereferencePushDown(Context context, JoinNode joinNode, Map<DereferenceExpression, Symbol> expressions, Assignments assignments)
+    {
+        List<Symbol> outputSymbols = joinNode.getOutputSymbols();
+        Map<Expression, Symbol> pushdownExpressions = expressions.entrySet().stream()
+                .filter(entry -> !isCastRowType(entry.getKey()))
+                .filter(entry -> outputSymbols.contains(getOnlyElement(extractAll(entry.getKey()))))
+                .collect(toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
+
+        if (pushdownExpressions.isEmpty()) {
+            return joinNode;
+        }
+
+        Map<Symbol, Expression> pushdownDereferences = pushdownExpressions.entrySet().stream().collect(toImmutableMap(Map.Entry::getValue, Map.Entry::getKey));
+        ImmutableMap.Builder<Symbol, Symbol> symbolsBuilder = ImmutableMap.builder();
+        for (Map.Entry<Expression, Symbol> entry : pushdownExpressions.entrySet()) {
+            symbolsBuilder.put(getOnlyElement(extractAll(entry.getKey())), entry.getValue());
+        }
+        Map<Symbol, Symbol> symbolsMap = symbolsBuilder.build();
+
+        List<EquiJoinClause> equiJoinClauses = joinNode.getCriteria().stream()
+                .map(EquiJoinClause::toExpression)
+                .map(expression -> replaceDereferences(expression, expressions))
+                .map(this::getEquiJoinClause)
+                .collect(toImmutableList());
```

Comment:
EqualJoinClause can only reference symbols; it cannot contain dereference expressions, can it?